ndcube/ndcube.py

```diff
@@ -444,42 +444,56 @@ def quantity(self):
         return u.Quantity(self.data, self.unit, copy=_NUMPY_COPY_IF_NEEDED)
 
     def _generate_world_coords(self, pixel_corners, wcs, needed_axes=None, *, units):
-        # Create meshgrid of all pixel coordinates.
-        # If user wants pixel_corners, set pixel values to pixel corners.
+        # This will generate only the coordinates that are needed if there is no correlation within the WCS
+        # This bypasses the entire rest of the function below which works out the full set of coordinates
+        # This only works for WCS that have the same number of world and pixel dimensions
+        if not pixel_corners and needed_axes is not None and not isinstance(wcs, ExtraCoords) and np.sum(wcs.axis_correlation_matrix[needed_axes]) == 1 and len(self.data.shape) == wcs.world_n_dim:
+            indices = [np.arange(self.data.shape[::-1][needed_axes[0]]) if wanted else [0] for wanted in wcs.axis_correlation_matrix[needed_axes][0]]
```

Comment:
I think something like below is what you need to make this valid of `pixel_corners` as well.
```suggestion
            lims = (-0.5, self.data.shape[::-1][needed_axes[0]] + 1) if pixel_corners else (0, self.data.shape[::-1][needed_axes[0]])
            indices = [np.arange(lims[0], lims[1]) if wanted else [0] for wanted in wcs.axis_correlation_matrix[needed_axes][0]]
```

Reply:
So I added this and the tests fail:

```
        world = wcs.pixel_to_world_values(0, 0, 0)
        if astropy.__version__ >= "7.0.0.dev":
            np.testing.assert_allclose(world, (14, np.nan, np.nan, -7.0))
            np.testing.assert_allclose(wcs.world_to_pixel_values(*world), (0, np.nan, np.nan))
        else:
>           np.testing.assert_allclose(world, (14, -12, -2.6e+10, -7.0))

../../.tox/py/lib/python3.13/site-packages/ndcube/wcs/wrappers/tests/test_compound_wcs.py:179: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = (<function assert_allclose.<locals>.compare at 0x7755ea22eac0>, array([14., nan, nan, -7.]), array([ 1.4e+01, -1.2e+01, -2.6e+10, -7.0e+00]))
kwds = {'equal_nan': True, 'err_msg': '', 'header': 'Not equal to tolerance rtol=1e-07, atol=0', 'strict': False, ...}

    @wraps(func)
    def inner(*args, **kwds):
        with self._recreate_cm():
>           return func(*args, **kwds)
E           AssertionError: 
E           Not equal to tolerance rtol=1e-07, atol=0
E           
E           nan location mismatch:
E            ACTUAL: array([14., nan, nan, -7.])
E            DESIRED: array([ 1.4e+01, -1.2e+01, -2.6e+10, -7.0e+00])
```
this code now seems to give the same output as astropy 7.0 will give in the future?!

Reply:
With this, I think it works perfectly thank you. I just have to account for a test change:

```
-> assert u.allclose(coords, [1.01e-09, 1.03e-09, 1.05e-09, 1.07e-09, 1.09e-09] * u.m)
(Pdb++) coords
CoordValues(em_wl=<Quantity [1.01e-09, 1.03e-09, 1.05e-09, 1.07e-09, 1.09e-09, 1.11e-09] m>)
(Pdb++) coords[0]
<Quantity [1.01e-09, 1.03e-09, 1.05e-09, 1.07e-09, 1.09e-09, 1.11e-09] m>
(Pdb++) 
```
I need to escape another dim which I didn't need to before?